Signup

sodeeqOlapade opened the pull request:
A pull request to merge signup feature-branch into develop branch
Test screenshot added.
<img width="758" alt="Screenshot 2019-08-07 at 12 53 56 PM" src="https://user-images.githubusercontent.com/48020768/62620923-74b3c880-b912-11e9-8131-52a0dca192fc.png">

Commit:
installed packages gravatar, jsonwebtoken and body-parser

Commit:
work on user validation and user model

add generateToken method on userSchema methods
add isMentor field to fields to be validated by Joi
add more fields(avatar) to userSchema

Commit:
implement the user sign up route

Commit:
installed jest and super test as dev dependencies

Commit:
wrote test for signup feature

Commit:
implement generateToken method to use the tokenEncoder method

Commit:
add token as arg to sendResponse function

include payload and token in the args for sendResponse function

rewrite test to test the newly added token and payload

Commit:
add signup endpoint to user.yaml

Tyav commented on controllers/user.controller.js:
```diff
@@ -1,25 +1,90 @@
+    }
+
+    //get user avatar
+    const avatar = gravatar.url(email, {
```
This can be used as a default value in the mongoose schema, 
```javascript
// user.model.js
avatar : {
    type: String,
    default : function(v = this.email) { return getAvatar(v))
}
// then in the helper, you could do something like this
// helper/avatar.js
module.exports = (email) => {
    gravatar.url(email, {
    // ...
})
}
```

Tyav commented on controllers/user.controller.js:
```diff
@@ -1,25 +1,90 @@
+    });
+
+    await user.save();
+    const fields = ['id', 'email', 'isAdmin', 'isMentor'];
```
why is this here? I think this has be made in the user model method, modify it there

Tyav commented on models/user.model.js:
```diff
@@ -61,32 +69,33 @@ UserSchema.methods = {
-    return pick(fields, this)
-  }, 
-  generateToken() {
+    const fields = ['id', 'name', 'email'];
```
The resembling logic in the [user.controller.js](https://github.com/Tyav/mentor-mentee/pull/4/files#r311534156) file  should be move or delete, modify this to suit your functionality and you are good.

Commit:
reimplement avatar field on user mode

creates avatar.js files which exports an avartarUrl generating function

default avatar field on user model to getAvatar's return value

Commit:
remove avatarUrl implementation

reimplement generateToken on user model to not take args and call it here

Commit:
rewrite the post to api/v1/users payload using jest toMatchObject
